### ISSUE TITLE
Array#to_formatted_s => List.to_formatted_s

### ISSUE THREAD

**Issue by mykewould** [mykewould/crutches#6]
Convert ActiveSupport [Array#to_formatted_s](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/array/conversions.rb#L89) to Crutches List.to_formatted_s

``` ruby
  # Extends <tt>Array#to_s</tt> to convert a collection of elements into a
  # comma separated id list if <tt>:db</tt> argument is given as the format.
  #
  #   Blog.all.to_formatted_s(:db) # => "1,2,3"
```

Hints to get started:
- Create a Crutches.List module (if it does not exist)
- Convert Ruby comments into Elixir [doctests](http://elixir-lang.org/getting-started/mix-otp/docs-tests-and-pipelines.html)
- TDD to solve
- [Additional tests can be found here](https://github.com/rails/rails/blob/master/activesupport/test/core_ext/array/conversions_test.rb#L79)

Additional resources:
[List intro](http://elixir-lang.org/getting-started/basic-types.html)
[List documentation](http://elixir-lang.org/docs/v1.0/elixir/List.html)

**Comment by rickr** [comment]
Hi Michael!

First off great job on this repo and its issues - it really does make it easy to jump into elixir as a newb. 

It looks like elixir's List.to_string is a bit different than ActiveSupport's - where in elixir it's assumed the list is integers representing a string and in ruby it just converts the entire array to a string: 

Examples:

```
iex(6)> List.to_string([101, 101])
"ee"
```

```
[28] pry(main)> [101, 101].to_s
=> "[101, 101]"
```

What mindset are you taking in these situations? Should this behave like you would expect it to in ruby or elixir? Did I miss the equivalent elixir implementation of Array#to_s?

**Comment by mykewould** [comment]
Thanks rickr! I'm a huge Rick and Morty fan, great name.

My mindset and approach are very open. I'd love everything to be discussed and worked on in the spirit of practice and learning. There's a lot of great thinking that's been floating around this repository, and I'd like to operate it by committee whenever possible. I'll kick this off with my thoughts, but I'd love to hear yours and others. If an agreement can't be made, I'll try to make a fair ruling.

I think the functionality displayed in ruby's to_s should exist in Crutches some place. My concern at this point would be naming. The obvious answer to me would be:

``` elixir
iex> List.to_s([101, 101])
"[101, 101]"
```

In summary, duplicate the functionality and the function name. The concern is that there may be another name that makes more sense within the context of Elixir/functional programming.

What are your thoughts? It'd be great to hear from other contributors as well! @knrz @druzn3k @janxious @duijf @nawns @philnash

**Comment by rickr** [comment]
My name is Rick so it resonates =)

FWIW I agree with you - my expectation as someone coming from ruby using an 'ActiveSupport' clone would be to see

```
iex> List.to_formatted_s([101, 101])
"[101, 101]"
```

Which is not at all what I would expect in elixir. 

My first thought was to implement a Crutches.List.to_s or something similar. Since this issue is extending ruby's Array#to_s it would be available for any other issues extending Array. It would also be accessible by users, causing them to explicitly demand ruby-ish behaviour. That then sets the tone of this lib as 'give me what I would expect in ruby'. Don't know if that's good or bad, right or wrong!

**Comment by mykewould** [comment]
It's a conundrum, because so much of Elixir and it's libraries borrow from Ruby in some way, it's hard to know when we've gone too far copying over behavior. It's good to know what your expectations would be, I think that gives us a good guideline. It's important for crutches to be consistent if nothing else. If we are going to follow a lot of Ruby naming and conventions this at least gives the user a paradigm to expect functionality with. I like the idea of creating a lib where you can accurately guess functionality by knowing the paradigm from which it was designed. I don't think we should knowingly go against the grain, or break any Elixir expectations to follow Ruby, but I can't see a reason not to follow Ruby in this case.

**Comment by druzn3k** [comment]
My 2c: `List.to_s` & `List.to_formatted_s` should collapse in a single method and only `List.to_formatted_s` should be exposed. In the original implementation, `to_formatted_s` [call already the original `to_s`](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/array/conversions.rb#L98). But I'm against creating and exposing a `List.to_s` method. The goal here is to reproduce the functionalities of ActiveSupport, not the Ruby language, and if we start to blur the boundaries from the start, we won't adapt, we are simply trying to reproduce _home_ in another place. If `to_s` does not work in the same way in Elixir, there should be a certain reason behind. :)

**Comment by CyrusNuevoDia** [comment]
With `to_formatted_s` acting like `to_s` unless you pass in the `:db` parameter, I think we should stick to how Elixir does it. I'm too new to Elixir to know if we can replicate the `respond_to?(:id)` behaviour that `to_formatted_s` calls when you pass `:db`, unless we check to see if a keyword list or a map has an `:id` key and use that.

`List.to_formatted_s` (or `List.to_formatted_string` imo, to keep with Elixir's naming of `List.to_string`) should fall back to Elixir's own `Kernel.to_string` (or `function unless passed`:db`. In which case it should look like this (for maps):

``` elixir
collection |> Enum.map(&Map.get(&1, :id)) |> to_string
```

**Comment by druzn3k** [comment]
@knrz the `:db` version should be built in unison with **Ecto**. Ecto's model leverage on structs for the implementation. In this case we could do something like this should work (oversimplified):

``` elixir
case elem
  %{id: obj_id} -> obj_id
  _ -> "null"
end
```

Why `Kernel.to_string`? Every base type has a `to_string` method, shouldn't we leverage on those instead on Kernel.to_string, that is built specifically to convert from characters list to string?

**Comment by druzn3k** [comment]
Ok, i checked the snippet and it seems to work:

```
iex(3)> defmodule User do
...(3)>   defstruct id: 10
...(3)> end
[...]
iex(7)> case %User{} do
...(7)>   %{id: obj_id} -> obj_id
...(7)>   _ -> "null"
...(7)> end
10
iex(8)> case %{nope: 0} do       
...(8)>   %{id: obj_id} -> obj_id
...(8)>   _ -> "null"            
...(8)> end
"null"
iex(9)> case 5 do                
...(9)>   %{id: obj_id} -> obj_id
...(9)>   _ -> "null"            
...(9)> end          
"null"
```

**Comment by janxious** [comment]
Second `List.to_formatted_string`. Second using Ecto's struct-based model usage.

**Comment by mykewould** [comment]
I like the idea of changing the name to `List.to_formatted_string` and implementing behavior that is more in line with Elixir. The expectations and results align better in my opinion. I think we should discuss how `.to_s` functionality should live in Elixir, and what it would look like, but it can be for another day/thread.

I also agree with using Ecto's struct-based model usage.

It sounds like we're all (mostly) in agreement?

@rickr do you agree?

**Comment by CyrusNuevoDia** [comment]
@druzn3k I can't seem to find `Map.to_string` or `Enum.to_string`. Tbh I'm sorta hitting out of my area when I said `Kernel.to_string`. Whatever you think is best.

**Comment by druzn3k** [comment]
@knrz you are right, I didn't check those! I searched and the only way I found for converting Maps & Lists to strings is `inspect`, but it doesn't _feel_ right. Damn...

:+1: for me to for changing the name to `to_formatted_string`!

**Comment by duijf** [comment]
I think the Elixir philosophy here is that when you convert between values, you want the semantics of the data to stay as much the same as they were. Converting `[101, 97]` to a binary with the `to_string` method gives you `"ea"`, which can be converted back to the original list with `to_char_list` (just a normal list, try `[101, 97] |> to_string |> to_char_list |> Kernel.++([0])` to see `[0]` appended).

I think the key point between the conversion functions is that there is no loss of information.

`inspect` seems to cover a different purpose to me. Not the conversion of values, but a way of presenting them (with the name `inspect` coming from the fact that it would mainly be used for debugging purposes).

For the above reason, I am wary of naming the function in the style of the `to_<something>` ones. Not sure how it should be called, but `format` seems better.

(As an aside, I am not convinced that this is any easier than `list |> Enum.join(",")`. It would become more interesting to me when we would provide ways of formatting different types from the standard library in the same vein, but I'm not sure how that would work.)

Up for debate as always :)

**Comment by druzn3k** [comment]
> For the above reason, I am wary of naming the function in the style of the `to_<something>` ones. Not sure how it should be called, but format seems better.

`humanized_join`?

> (As an aside, I am not convinced that this is any easier than list |> Enum.join(","). It would become more interesting to me when we would provide ways of formatting different types from the standard library in the same vein, but I'm not sure how that would work.)

The purpose of this function is not to output a joined array, but an array joined in a _human_ way, so `[1,2,3]` became `"1, 2 and 3"` and not `"1, 2, 3"`. We could always implement [protocols](http://elixir-lang.org/getting-started/protocols.html) as suggested in another issue if we want to provide formatting primitives.

**Comment by duijf** [comment]
> The purpose of this function is not to output a joined array, but an array joined in a human way, so [1,2,3] became "1, 2 and 3" and not "1, 2, 3". We could always implement protocols as suggested in another issue if we want to provide formatting primitives.

@druzn3k I don't think so. Wasn't that `to_sentence` (#5, already implemented)?

Anyway, I've got some conflicting views (`to_sentence` seems better to me than `to_formatted_string`, but I'm not sure why), so I've got some thinking to do.

**Comment by druzn3k** [comment]
LOL, you are right @duijf, I got it confused! :+1:

**Comment by rickr** [comment]
Hello all, 

Sorry for the delay - I was out of town this weekend.

@duijf brings up a good point. I'll do some research but is anyone aware of any instances of to_$FOO methods that already exist and are destructive? Could we note that this method is destructive in the documentation? 

I think renaming would be a bad idea if we're going for parity we ActiveSupport. If we want to just be inspired by it that sounds like a good solution.

**Comment by druzn3k** [comment]
The goal is to became quickly ActiveSupport inspired. It's unavoidable, but for now we are still heavily using ActiveSupport as a base, until it will be good/large _enough_ to diverge (but I think that with the last activities we are already moving in that direction). So this could became the _turning point_ in which we start officially to diverge. :+1:

**Comment by druzn3k** [comment]
closing this in favour of reopening a new issue less ambiguous
